platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorSearch.java

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.consensus;
+
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+/** Used for searching for ancestors events in a hashgraph */
+public class AncestorSearch {
+    /** the iterator that returns ancestors */
+    private final AncestorIterator iterator;
+
+    public AncestorSearch() {
+        this(new EventVisitedMark());
+    }
+
+    /**
+     * Create a new ancestor search with the given mark
+     * @param mark the instance to use to mark events as visited
+     */
+    public AncestorSearch(@NonNull final EventVisitedMark mark) {
+        this.iterator = new AncestorIterator(mark);
+    }
+
+    /**
+     * Gets an iterator over the set of ancestors of the given event (the root of the search) for which valid is true.
+     * This will not include a valid ancestor that is only reachable through invalid ancestors.
+     *
+     * @param root the root event whose ancestors should be searched
+     * @param valid do a depth-first search, but backtrack from any event e where valid(e)==false
+     * @return an iterator over all valid ancestors of the root
+     */
+    public @NonNull AncestorIterator search(@NonNull final EventImpl root, @NonNull final Predicate<EventImpl> valid) {
+        iterator.search(root, valid);
+        return iterator;
+    }
+
+    /**
+     * Finds events that are ancestors to all the supplied events. It will only traverse valid events,
+     * where valid is defined by the predicate supplied. This method will also populate {@link
+     * EventImpl#getRecTimes()} for the events returned.
+     *
+     * @param events the events whose ancestors we are searching for
+     * @param valid checks if the event should be part of the search
+     * @return a list of all common ancestors
+     */
+    public @NonNull List<EventImpl> commonAncestorsOf(
+            @NonNull final List<EventImpl> events, @NonNull final Predicate<EventImpl> valid) {
+        // each event visited by iterator from at least one of the provided events
+        final ArrayList<EventImpl> visited = new ArrayList<>();
+        // Do a non-recursive search of the hashgraph, without using the Java stack, and being
+        // efficient when it's a DAG that isn't a tree.
+        for (final EventImpl e : events) {
+            final AncestorIterator validAncestors = search(e, valid);
+            while (validAncestors.hasNext()) {
+                final EventImpl event = validAncestors.next();
+                if (event.getRecTimes() == null) {
+                    event.setRecTimes(new ArrayList<>());
+                    visited.add(event);
+                }
+                event.getRecTimes().add(validAncestors.getTime());
```

Comment:
Why is this method updating the events? Either this updating should be extracted, or the javadocs should be very clear about what is being updated.

Reply:
This is needed for determining the consensus timestamps of events. This is old code, repackaged. The javadoc indicates that this is populated